gtests/pk11_gtest/pk11_kbkdf.cc

```diff
@@ -0,0 +1,72 @@
+/* -*- Mode: C++; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
+/* vim: set ts=2 et sw=2 tw=80: */
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include <memory>
+#include "nss.h"
+#include "pk11pub.h"
+#include "secerr.h"
+#include "sechash.h"
+
+#include "blapi.h"
+
+#include "gtest/gtest.h"
+#include "nss_scoped_ptrs.h"
+#include "util.h"
+
+#define DIM(a) (sizeof((a))/sizeof((a)[0]))
+
+namespace nss_test {
+class Pkcs11KbkdfTest : public ::testing::Test {
+ protected:
+  ScopedPK11SymKey ImportKey(CK_MECHANISM_TYPE mech, SECItem *key_item) {
+    ScopedPK11SlotInfo slot(PK11_GetInternalSlot());
+    if (!slot) {
+      ADD_FAILURE() << "Can't get slot";
+      return nullptr;
+    }
+
+    ScopedPK11SymKey result(PK11_ImportSymKey(
+        slot.get(), mech, PK11_OriginUnwrap, CKA_SIGN, key_item, nullptr));
+
+    return result;
+  }
+
+  void RunTest(void) {
+    uint8_t key[AES_128_KEY_LENGTH] = {0x2B, 0x7E, 0x15, 0x16, 0x28, 0xAE,
+                                       0xD2, 0xA6, 0xAB, 0xF7, 0x15, 0x88,
+                                       0x09, 0xCF, 0x4F, 0x3C};
+
+    SECItem key_item = {siBuffer, key, sizeof(key)/sizeof(key[0])};
+    ScopedPK11SymKey p11_key = ImportKey(CKM_AES_CMAC, &key_item);
+
+    CK_SP800_108_COUNTER_FORMAT iterator = {CK_FALSE, 8};
+    CK_SP800_108_DKM_LENGTH_FORMAT dkm = { CK_SP800_108_DKM_LENGTH_SUM_OF_KEYS, CK_FALSE, 8 };
+
+    CK_PRF_DATA_PARAM dataParams[] = {
+      { CK_SP800_108_ITERATION_VARIABLE, &iterator, sizeof(iterator) },
+      { CK_SP800_108_DKM_LENGTH, &dkm, sizeof(dkm) }
+    };
+
+    CK_SP800_108_KDF_PARAMS kdfParams =
+    {
+      CKM_AES_CMAC,
+      2,
+      dataParams,
+      0,       /* no additional derived keys */
+      NULL     /* no additional derived keys */
+    };
```

Comment:
My intent is to add everything from NIST's published [CAVP tests](https://csrc.nist.gov/Projects/Cryptographic-Algorithm-Validation-Program/Key-Derivation) for this, since this implementation _should_ be able to support it all.

Reply:
@rjrelyea I've since added all CAVP tests for Counter mode, sans the TDES / 2TDES tests. So this includes:

 - AES-128
 - AES-192
 - AES-256
 - HMAC-SHA1
 - HMAC-SHA224
 - HMAC-SHA256
 - HMAC-SHA384
 - HMAC-SHA512

So there should be no surprises here during FIPS certification of this. OTOH, there are a _lot_ of tests:

 - Fixed data before ("counter after" / `AFTER_FIXED`)
 - Fixed data before and after ("counter middle" / `MIDDLE_FIXED`)
 - Fixed data after ("counter before" / `BEFORE_FIXED`)

With `r` bitlen between:

 - 8
 - 16
 - 24
 - 32

Each of these has several tests for each of the different ciphers. My thought was to prune this to say, 1-2 of each (`r` bit length, MAC function, output length) combination. Thoughts?